Add docs page on deriving array rules

sethaxen opened the pull request:
At @oxinabox's request, I wrote up a (very, very rough) first draft of the modification of Giles' method that I find useful for deriving forward- and reverse-mode rules of functions of arrays. I'd appreciate some feedback on the scope, flow, and content of this early version.

nickrobinson251 commented:
Awesome. Will look forward to reading this.

Link to preview of the docs: http://www.juliadiff.org/ChainRulesCore.jl/preview-PR191/arrays.html

sethaxen commented:
Yikes `\overline{A}^H` looks ugly in MathTeX.

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+
+$$dC\ A + C\ dA = 0$$
+$$dC\ A\ C + C\ dA\ C = 0$$
+$$dC = -C\ dA\ C$$
```
I think you meant to invert both occurrences of `C` in this line.

sethaxen replied:
Oh, the slashes actually aren't characters, they add more spacing between the characters, since otherwise it's tricky to see that `dA` is one unit. Perhaps there's a way to do that that is friendlier to read the source of.

oxinabox replied:
Perhaps `~` which is nonbreaking space, or `\;`,  `\quad` etc?

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+
+### Matrix inversion
+
+$$C A = A C = I$$
```
For the sake of flow, maybe start by stating `C := inv(A)` or whatever to make it clear what function the rule is being derived for. When you're reading these derivations for the first time I think it's helpful to make it clear what the input is and what the output is so that you don't get lost.

simeonschaub replied:
What do you think about just using `A^{-1}` instead of `C`?

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+
+where $\operatorname{tr}$ is the `trace`.
+
+Plugging in the identity for $dC$, we get
```
```suggestion
Plugging in the expression for $dC$, we get
```
On first reading I thought you meant the identity _matrix_.

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+# Arrays
+
+When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles' method is a quick way of deriving `frule`s and `rrule`s.
+Giles' method is succinctly explained in [this paper](https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf), but we will slightly generalize it.
```
Perhaps comment on why it's been generalised, otherwise the reader won't discover it until they read your `!!!note` further down.

sethaxen replied:
Good idea, I'll introduce the comparison of the note at the top.

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+
+By applying the same definition of $ds$ to the intermediates $A$ and $B$, we get
+
+$$ds = \Re\left(\operatorname{tr}(\overline{A}^H\ dA)\right) + \Re\left(\operatorname{tr}(\overline{B}^H\ dB)\right)$$
```
I think another line here that explicitly highlights the fact that you've equated `\overline{A}^H` and ``\overline{C}^H df / dA` would be helpful. There's a lot of notation flying around here, so it might well take someone who's new to this a little while to figure out that you're just equating terms.

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+$$ds = \Re\left(\operatorname{tr}(B \overline{C}^H dA)\right) + \Re\left(\operatorname{tr}(\overline{C}^H A\ dB)\right)$$
+
+That's it!
+The expression is in the desired form to solve for the adjoints:
```
Perhaps hold the reader's hand a little here by saying something like that you're matching terms in the above equation.

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+$$\overline{A} = (B \overline{C}^H)^H = \overline{C} B^H$$
+$$\overline{B} = (\overline{C}^H A)^H = A^H \overline{C}$$
+
+Using our notation conventions, we would implement the `rrule` as
```
Consider `Using ChainRules' notation`, or something like that to distinguish from the notation in this doc.

willtebbutt commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+
+Again, we derived the differential identity as
+
+$$dC = -C dA C$$
```
inv again :)

willtebbutt commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,219 @@
+Because none of this derivation really depended on the index (or indices), we can easily implement the `rrule` more generically using broadcasting:
+
+```julia
+function rrule(::typeof(sum), ::typeof(abs2), A::Array; dims = :)
````
```suggestion
function rrule(::typeof(sum), ::typeof(abs2), A::Array{ComplexF64}; dims = :)
```

sethaxen replied:
These rules should work for any `Real` and `Complex`, so any objection to `Union{Real,Complex}`?

willtebbutt commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,219 @@
+We can implement the resulting `rrule` as:
+
+```julia
+function rrule(::typeof(inv), A::Matrix)
````
```suggestion
function rrule(::typeof(inv), A::Matrix{Float64})
```

willtebbutt commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,219 @@
+Using our notation conventions, we would implement the `rrule` as
+
+```julia
+function rrule(::typeof(*), A::Matrix, B::Matrix)
````
```suggestion
function rrule(::typeof(*), A::Matrix{Float64}, B::Matrix{Float64})
```

willtebbutt commented:
Once we've done a couple of iterations I'll put out a review request in the AD channel on Julia Slack / Zulip -- it would be great to get a couple of mathematically inclined people who've not done much with AD before to take a look as well.

oxinabox commented on docs/make.jl:
```diff
@@ -29,6 +29,7 @@ makedocs(
         "FAQ" => "FAQ.md",
         "Writing Good Rules" => "writing_good_rules.md",
         "Complex Numbers" => "complex.md",
+        "Arrays" => "arrays.md",
```
perhaps we should term this "Deriving Array Rules" or something of the like?

simeonschaub commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,219 @@
+
+But this is just a sum matrix products:
+
+$$C = dA\ B + A\ dB$$
```
```suggestion
$$dC = dA\ B + A\ dB$$
```

sethaxen commented:
I've implemented some (not all yet) of the recommended changes.
I also changed the notation to look more like Julia functions (e.g. `real` instead of `ℜ`, `conj(x)` instead of `x^*`; `adjoint(x)` is still `x^H` to avoid confusion).
And for multi-step derivations, I added step annotations with equation references (still some more cleaning to do here).
Still need to make the introduction of reverse-mode easier to follow.

sethaxen commented:
Hm, though it looks nice, maybe the step annotation and equation linking is not the way to go. Those are only supported by MathJax. Recent versions of Documenter (latest is v0.25, but we use v0.23) default to KaTeX, which renders faster, but can optionally still use MathJax. I don't have a strong opinion either way.

ettersi commented:
It is unclear to me whether the aim of this write-up is to explain general recipes for deriving forward and backward rules, or whether the aim is to assemble a list of particular rules together with their proofs. In the former case, I think it would be good to spend a bit more time explaining the method. For example, it is unclear to me what is the difference between `dA` and `\dot A`.

sethaxen commented:
> It is unclear to me whether the aim of this write-up is to explain general recipes for deriving forward and backward rules, or whether the aim is to assemble a list of particular rules together with their proofs.

Definitely the former. The goal is to give users who've never derived their own rules a recipe for how to do that in both modes, with some examples to demonstrate. I'll clarify that in the intro.

> In the former case, I think it would be good to spend a bit more time explaining the method. For example, it is unclear to me what is the difference between `dA` and `\dot A`.

Good point. The hand-wavy explanation is that you get `\dot A` by dividing `dA` by the differential `dt`, which is not technically true, but like treating derivatives as ratios of differentials is useful, it's a notational convenience that works when one wants to keep things simple. I definitely want to tighten up the mathematical justification without getting much more theoretical.

ettersi commented:
As far as I am aware, there is no universally accepted definition of differentials. In an equation like `dC = dA + dB`, I like to think of all the `d` quantities as derivatives with respect to an unspecified scalar variable, hence `dC = dA + dB` and `\dot C = \dot A + \dot B` are essentially the same to me. Maybe it would be better to avoid abstract differentials to avoid the risk of such misinterpretations?

simeonschaub commented:
They do have quite a concise and universal definition as [differential one-forms](https://en.wikipedia.org/wiki/Differential_form#Differential_calculus). I do think as the covariant derivative, they are more intuitive for reverse mode / pullbacks though, so maybe we should try to be consistent and only use the dot notation for forward mode.

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+# Deriving Array Rules
+
+When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles' method is a quick way of deriving [`frule`](@ref)s and [`rrule`](@ref)s.
```
Usually the second `s` is added if it would be pronouned (which i think it would be for "Giles's method"

```suggestion
When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles's method is a quick way of deriving [`frule`](@ref)s and [`rrule`](@ref)s.
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+# Deriving Array Rules
+
+When the inputs and outputs of the functions are arrays (potentially with scalars), a modified version of Giles' method is a quick way of deriving [`frule`](@ref)s and [`rrule`](@ref)s.
+Giles' method is succinctly explained in [^Giles2008] and its extended work [^Giles2008ext], but we will generalize it to handle arrays with both real and complex entries and arrays of arbitrary dimensions.
```
```suggestion
Giles's method is succinctly explained in [^Giles2008] and its extended work [^Giles2008ext], but we will generalize it to handle arrays with both real and complex entries and arrays of arbitrary dimensions.
```

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,440 @@
+\end{align*}
+```
+
+Giles' method for deriving pullback functions is to first derive the differential identity (i.e. pushforward) using the above approach, then pre-multiply by $\overline{C}^H$, and take the real trace.
````
```suggestion
Giles's method for deriving pullback functions is to first derive the differential identity (i.e. pushforward) using the above approach, then pre-multiply by $\overline{C}^H$, and take the real trace.
```

nickrobinson251 replied:
I wonder if this paragraph and the properties of the trace function should come before the example in which they are used? Or perhaps for this little section to have a sub-heading? It feels a bit lost amongst the text right now.

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,440 @@
+```
+
+!!! note
+    Our method is identical to Giles' approach, except we have replace the transpose with the conjugate transpose, and we have added the constraint that the inner product be real.
````
```suggestion
    Our method is identical to Giles's method, except we have replace the transpose with the conjugate transpose, and we have added the constraint that the inner product be real.
```

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,440 @@
+
+```math
+\begin{align*}
+    dC_{i1k} &= \sum_j \operatorname{real}( \operatorname{conj}(dA_{ijk})~ A_{ijk} +
````
```suggestion
    dC_{i1k} &= \sum_j \operatorname{real}\left( \operatorname{conj}(dA_{ijk})~ A_{ijk} +
```

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,440 @@
+```math
+\begin{align*}
+    dC_{i1k} &= \sum_j \operatorname{real}( \operatorname{conj}(dA_{ijk})~ A_{ijk} +
+                            \operatorname{conj}(A_{ijk}) ~dA_{ijk} ) \\
````
```suggestion
                            \operatorname{conj}(A_{ijk}) ~dA_{ijk} \right) \\
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+\begin{align*}
+    dC_{i1k} &= \sum_j \operatorname{real}( \operatorname{conj}(dA_{ijk})~ A_{ijk} +
+                            \operatorname{conj}(A_{ijk}) ~dA_{ijk} ) \\
+             &= \sum_j \operatorname{real}(
```
```suggestion
             &= \sum_j \operatorname{real}\left(
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}
+                    \right) +
+                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}
+                )\\
```
```suggestion
                \right\\
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+                    \right) +
+                        \operatorname{conj}(A_{ijk}) ~dA_{ijk}
+                )\\
+             &= \sum_j 2 \operatorname{real}( \operatorname{conj}(A_{ijk}) ~dA_{ijk} )
```
```suggestion
             &= \sum_j 2 \operatorname{real}\left( \operatorname{conj}(A_{ijk}) ~dA_{ijk} \right)
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+function frule(
+    (_, _, ΔA),
+    ::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex};
+    dims = :,
```
should `dims=2`? since the primal function [shown above](https://github.com/JuliaDiff/ChainRulesCore.jl/pull/191/files#diff-04495d68483dba4eac9bf4246b9c62d0R342) has `dims=2`?

```suggestion
    dims=2,
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+    ::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex};
+    dims = :,
+)
+    C = sum(abs2, A; dims = dims)
```
```suggestion
    C = sum(abs2, A; dims=dims)
```

sethaxen replied:
I ran the equations through JuliaFormatter, so I'm happy with the current spacing.

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+    dims = :,
+)
+    C = sum(abs2, A; dims = dims)
+    ∂C = sum(2 .* real.(conj.(A) .* ΔA); dims = dims)
```
```suggestion
    ∂C = sum(2 .* real.(conj.(A) .* ΔA); dims=dims)
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+\begin{align*}
+    ds &= \operatorname{real} \left(\sum_{ijk}
+                  \operatorname{conj}(\overline{C}_{i1k})
+                  2 \operatorname{real}(\operatorname{conj}(A_{ijk}) ~dA_{ijk})
```
```suggestion
                  2 \operatorname{real}\left( \operatorname{conj}(A_{ijk}) ~dA_{ijk} \right)
```

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,440 @@
+We now solve for $\overline{A}$:
+
+```math
+\overline{A}_{ijk} = \operatorname{conj}(
````
```suggestion
\overline{A}_{ijk} = \operatorname{conj}\left(
```

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+\overline{A}_{ijk} = \operatorname{conj}(
+                         2 \operatorname{real}( \overline{C}_{i1k} )
+                         \operatorname{conj}(A_{ijk})
+                     )
```
```suggestion
                     \right)
```

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,440 @@
+Because none of this derivation really depended on the index (or indices), we can easily implement the `rrule` more generically using broadcasting:
+
+```julia
+function rrule(::typeof(sum), ::typeof(abs2), A::Array{<:RealOrComplex}; dims = :)
````
same question re `dims`?

sethaxen replied:
In this case I wanted to derive a generic rule but writing that out in math notation is horrific, so I instead derived for an arbitrary choice of indices and then stated that it's generic. I added some more text (hopefully) making that clearer.

nickrobinson251 commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,440 @@
+        ∂A = @thunk(2 .* real.(ΔC) .* A)
+        return (NO_FIELDS, ∂abs2, ∂A)
+    end
+    return sum(abs2, A; dims = dims), sum_abs2_pullback
```
```suggestion
    return sum(abs2, A; dims=dims), sum_abs2_pullback
```

sethaxen commented:
Thanks everyone for the feedback! I did quite a substantial refactor to introduce the approach with generic arrays first and eliminate abstract differentials.

sethaxen commented:
Okay, looks good to me. I know a few folks already approved it (thanks!) but since I rewrote a lot, I'd appreciate 1 or 2 people giving it another look-over. Also @willtebbutt it should be ready to share on Slack and/or Zulip now.

simeonschaub commented on docs/src/arrays.md:
```diff
@@ -0,0 +1,562 @@
+
+$$t \mapsto (\ldots, X_m, \ldots) \mapsto \Omega \mapsto s,$$
+
+At any step in the program, if we have intermediates $X_m$, we can write down the derivative $\frac{ds}{dt}$ in terms of the tangents $\dot{X}_m = \frac{dX_m}{dt}$ and adjoints $\overline{X}_m = \frac{\partial s}{\partial X_m}$
```
Should `$\overline{X}_m$` be superscript, or do you think that would be too confusing for the targeted audience?

sethaxen replied:
I considered it, but yeah, I think it would be too confusing. (this tutorial really should be using Einstein notation, which would really simplify the equations, but I also ruled that out for the same reason).

simeonschaub replied:
Yeah, seems reasonable. Probably also hard to keep this consistent throughout the whole tutorial. I don't think we can use Einstein notation here anyways, since sometimes twice-occuring indices are not summed over.

sethaxen commented:
Okay at @wupeifan's suggestion (thanks!) I added another section to the end deriving the rules for `logabsdet`, since it returns a tuple of outputs.

sethaxen commented:
I added a note on checking that pushforwards and pullbacks are sensible, because it's a cool example for complex sign.
Unless there are any objections, I'll merge this PR tomorrow.

willtebbutt commented:
Thanks for the great work @sethaxen . Please feel free to merge whenever you fancy.

nickrobinson251 commented on docs/src/arrays.md:
````diff
@@ -0,0 +1,800 @@
+\end{align*}
+```
+
+!!! note
````
tiny things, but i wonder is this note can be moved to after the `rrule` code below? Since it discusses both pushforwards and pullbacks, it kinda breaks the flow of the section on Reverse-mode rule with where it is now

sethaxen replied:
Good point, thanks. Done!
